### DISCUSSION TITLE
CommunicationsException causing Mirth slowdown requiring Mirth restart

### DISCUSSION THREAD

**Discussion by BuddhaBuddy1** [nextgenhealthcare/connect#5834] [Q&A]
We sometimes get "com.mysql.cj.jdbc.exceptions.CommunicationsException: Communications link failure" from a connection we got from DatabaseConnectionFactory.createDatabaseConnection(). When that happens a few times, Mirth sometimes starts to get slow, and we have to restart it to recover from the backlog. The log shows that it is happening on our script code, not in Mirth code.

The next line in the exception either says something like this:

> The last packet successfully received from the server was 30,307 milliseconds ago.  The last packet sent successfully to the server was 30,307 milliseconds ago.

... or this:

> The last packet sent successfully to the server was 0 milliseconds ago. The driver has not received any packets from the server.

From what I read, people say this means that you have a bad connection in the connection pool, and you should add a validationQuery to your pool configuration, but as far as I can tell, we don't HAVE a connection pool. We call createDatabaseConnection() every time. DatabaseConnectionFactory seems to be a pretty thin wrapper over a DatabaseConnection, so I don't really get what is happening here.

What I THINK is going on here is that the exception in our script isn't really the problem, it is just showing us that there IS a problem. However, mirth.log doesn't show that any exceptions are happening there. I just don't see how a few exceptions in our script can cause our server to run slower. In fact, there is NOTHING in mirth.log but our exceptions and a few startup and shutdown messages. There isn't some secret mirth log somewhere other than /logs, is there? :-)

Any ideas? I'll probably try to switch to using a connection pool, but I just don't see what could be causing a slowdown here.

I don't KNOW that these exceptions are the cause of (or a symptom of) the slowdown, but we don't get a backlog unless we have these exceptions, although we don't always get a backlog if we see these exceptions.

**Comment by pacmano1** [comment]
I would stop creating a connection every time.  And create a shared dbConn in the globalMap as a first step.  

But any other errors in the DB log?

**Comment by tonygermano** [comment]
Are you still closing the connections when errors occur? Are you connecting to the mirth backend db or some other db?

**Reply by pacmano1** [reply]
I should have said using a cached connection would by default not open a new connection on every call and thereby address non closed connections.  

e.g.:

https://github.com/pacmano1/Mirth-Snippets/blob/main/executeSharedDBStatement.js

**Comment by BuddhaBuddy1** [comment]
@pacmano1, no other errors, just this one, and it happens anywhere from 4 to 12 times a day, usually in the mornings.

@tonygermano, we are contacting our own database, although in the same server as the mirth ones. I just wasn't sure if this error was known to happen WITHOUT a connection pool. I'm not sure I'd trust a shared connection like in your example. Does mirth process everything single threaded? We have multiple inbound channels.

I think we just need to start using a connection pool.

**Reply by pacmano1** [reply]
What is your message volume?  Number of channels?   There is nothing that precludes you from usng a gboal dbConn per channel.   I am suggessting openning and closing a connection every time is not the best practice, usally due to programmer error.

I don't know what your definision of "trust" is, that code came from the advanced mirth class (I think) and has been in use for 9 years on many engines that get tens of thousands of messages a day.

If you are getting millions of message a day then for sure I would be looking at some other pool.

**Comment by BuddhaBuddy1** [comment]
So "globalMap" is per channel, or would it need to be stored elsewhere for a per-channel connection?
Sorry, I'm not the Mirth guy, he conveniently left the country on vacation this week. :-D

[edit] Oh, and we are not in the millions, but around 1k per hour.

**Reply by pacmano1** [reply]
A globalMap (aka $g) is accoss the entire engine.  And at 1k messages an hour I would not expect much of a problem with a single shared connection accros the entire engine.   If you are worred about it, you can change the code template to create a dbConn for a parameter passed to the function.     Lots of options.  The point is not to open and close connections every time.
